### ISSUE TITLE
While running a developement chain, execution of Solidity assembly code throws

### ISSUE THREAD

**Issue by fratellopartigiano** [paritytech/parity#8899]
> I'm running:
>
> - **Which Parity version?**: Parity v1.11.1 (tried also 1.10.4)
> - **Which operating system?**: Ubuntu 16.04.4 LTS, kernrel 4.4.0-127-generic
> - **How installed?**: docker, binaries
> - **Are you fully synchronized?**: irrelevant
> - **Which network are you connected to?**: devchain, kovan
> - **Did you try to restart the node?**: irrelevant

> In addition:
> - Solidity 0.4.24 (tried also 0.4.23)

**Might be related to #8898.**
Avoiding the issue mentioned above by using `web3py` (or a more sophisticated `.js` script), I tried to run a proxy smart contract (written in Solidity using assembly `delegatecall`). **Although it works fine on `kovan` or in `remix`**, it always failed on a development chain (I've tried custom PoW and PoA chainspecs, as well as a copy of kovan chainspec).

Calling the `fallback` function in `Proxy` contract (with verified proper logic implementation already assigned) with the signature of `doSomething()` function, always throws on a development chain. It was not possible to figure out which line of the assembly code caused the problem (commenting out other lines led to no conclusion).

I've run that on 2 different machines, using dockerized and undockerized environment, always with the same result.

The used smart contract code are the following (python deployment code too complex too include):

*Proxy.sol*
```
pragma solidity ^0.4.24;

// taken from openzeppelin and edited

/**
 * @title Proxy
 * @dev Gives the possibility to delegate any call to a foreign implementation.
 */
contract Proxy {

    address public impl;

    constructor() public {
    }

    function upgradeTo(address newimpl) external onlyOwner {
	impl = newimpl;
    }
    
    function implementation() public view returns (address) {
	return impl;
    }

    function () payable public {
	address _impl = implementation();
	require(_impl != address(0));

	assembly {
	    let ptr := mload(0x40)
	    calldatacopy(ptr, 0, calldatasize)
	    let result := delegatecall(gas, _impl, ptr, calldatasize, 0, 0)
	    let size := returndatasize
	    returndatacopy(ptr, 0, size)

	    switch result
	    case 0 { revert(ptr, size) }
	    default { return(ptr, size) }
	}
    }
}
```

*LogicPrimary.sol*
```
pragma solidity ^0.4.24;

contract LogicPrimary {

    address public impl;
    
    constructor() public {
    }

    function doSomething() public pure returns(uint32) {
	return 1;
    }
}
```

**Comment by Tbaut** [comment]
What development chain did you use exactly? Please also upgrade to 1.11.3

**Comment by fratellopartigiano** [comment]
Here is the chainspec I was using (in addition, I tried also with an exact copy of kovan chainspec with just modified validators so that I could use it privately).

```
{
    "name": "DevelopmentChain1",
    "engine": {
	"authorityRound": {
	    "params": {
		"gasLimitBoundDivisor": "0x400",
		"registrar" : "0xfAb104398BBefbd47752E7702D9fE23047E1Bca3",
		"stepDuration": "1",
		"blockReward": "0x4563918244F40000",
		"validateScoreTransition": 1000000,
		"eip155Transition": 1000000,
		"validateStepTransition": 1500000,
		"validators": {
		    "list": ["0x007CcfFb7916F37F7AEEf05E8096ecFbe55AFc2f"]
		}
	    }
	}
    },
    "params": {
	"gasLimitBoundDivisor": "0x400",
	"maximumExtraDataSize": "0x20",
	"minGasLimit": "0x13880",
	"networkID" : "0x2AA",
	"validateReceiptsTransition" : 1000000
    },
    "genesis": {
	"seal": {
	    "authorityRound": {
		"step": "0x0",
		"signature": "0x0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
	    }
	},
	"difficulty": "0x20000",
	"gasLimit": "0x5F5E100"
    },
    "accounts": {
	"0000000000000000000000000000000000000001": { "balance": "1", "builtin": { "name": "ecrecover", "pricing": { "linear": { "base": 3000, "word": 0 } } } },
	"0000000000000000000000000000000000000002": { "balance": "1", "builtin": { "name": "sha256", "pricing": { "linear": { "base": 60, "word": 12 } } } },
	"0000000000000000000000000000000000000003": { "balance": "1", "builtin": { "name": "ripemd160", "pricing": { "linear": { "base": 600, "word": 120 } } } },
	"0000000000000000000000000000000000000004": { "balance": "1", "builtin": { "name": "identity", "pricing": { "linear": { "base": 15, "word": 3 } } } },
	"0x007ccffb7916f37f7aeef05e8096ecfbe55afc2f": {
	    "balance": "20000000000000000000"
	},
	"0x99429f64cf4d5837620dcc293c1a537d58729b68": {
	    "balance": "20000000000000000000"
	},
	"0xca247d7425a29c6645fa991f9151f994a830882d": {
	    "balance": "20000000000000000000"
	},
	"0x794f74c8916310d6a0009bb8a43a5acab59a58ad": {
	    "balance": "20000000000000000000"
	},
	"0x276ecb88715a503b00d1f15af4c17dc051991667": {
	    "balance": "20000000000000000000"
	},
	"0x83042c0147acce98e35ed9ef52e6dfc5c67ef92e": {
	    "balance": "20000000000000000000"
	},
	"0x8ab7114ba0f7ca706af69f799588766c8426aa24": {
	    "balance": "20000000000000000000"
	},
	"0x932d9e95e5d2cac02eebbe6763ab2c7b0a9d6a2f": {
	    "balance": "20000000000000000000"
	},
	"0x893c3f80d2a0375b3f00f856cf8a6775e4efc26a": {
	    "balance": "20000000000000000000"
	}
    }
}
```

**Comment by fratellopartigiano** [comment]
Looks like I've solved it, I was missing the transition blocks for the Byzantium hard fork in the chainspec. At least made the kovan-clone working now when I changed the block numbers to 1.

**Comment by fratellopartigiano** [comment]
I can confirm that the above chainspec also works after adding
```
"eip155Transition": 1,
"maxCodeSize": 24576,
"maxCodeSizeTransition": 1,
"validateChainIdTransition": 1,
"validateReceiptsTransition" : 1,
"eip140Transition": 1,
"eip211Transition": 1,
"eip214Transition": 1,
"eip658Transition": 1,
"wasmActivationTransition": 1
```
to `params` and setting the existing transition blocks to 1. This also solved the #8898.

Thanks and apologies for confusion. I hope it may be useful for someone else in the future.
